File: game/crafting.py
"""Book-defined item improvements with a single reversible material transaction."""
import copy
from django.core.exceptions import PermissionDenied
from django.shortcuts import get_object_or_404
from .models import Entry, Item
from .rules import Change, current_scene
from .inventory import amount, new_item
from . import enchantments, weaponry

ELEMENTS={'Малое изначальное зачарование':'Изначальная',**{'Малое зачарование '+suffix:name for suffix,name in [('огня','Огонь'),('воды','Вода'),('земли','Земля'),('воздуха','Воздух'),('тьмы','Тьма'),('света','Свет'),('молнии','Молния'),('холода','Холод'),('природы','Природа')]}}
ALCHEMY={'Малое зелье лечения':'healing_potion','Малый яд':'poison','Масло точности':'accuracy_oil'}


def profile(entry):
    if entry.name in ALCHEMY:return {'craft':'Алхимик','kind':'production','product':ALCHEMY[entry.name],'hours':3}
    if enchantments.profile(entry):return {'craft':'Зачарователь','kind':'enchantment','dust':50,'element':ELEMENTS.get(entry.name,entry.data.get('element',''))}
    if weaponry.profile(entry):return {'craft':'Оружейник','kind':'upgrade',**weaponry.profile(entry)}


def catalogue():
    return [{'id':e.id,'name':e.name,**profile(e)} for e in Entry.objects.filter(archived=False) if profile(e)]


def apply(user,p):
    from .views import owned, access_campaign
    c=owned(user,p['character'])
    if current_scene(c):raise ValueError('Мастерская доступна вне боя')
    recipe=get_object_or_404(Entry,pk=p['recipe'],archived=False)
    spec=profile(recipe)
    if not spec:raise ValueError('Выберите рецепт улучшения')
    known=c.abilities.filter(pk=recipe.id).exists() or c.knowledge.filter(entry=recipe,kind='recipe',archived=False,**({} if c.owner_id==user.id else {'private':False})).exists()
    crafts=set(Entry.objects.filter(pk__in=c.info.get('craft_ids',[]),kind='craft').values_list('name',flat=True))
    crafts.add(str(c.info.get('craft','')))
    if not known or spec['craft'] not in crafts:raise ValueError('Выберите нужное ремесло и изучите рецепт')
    def stock(pk,revision):
        item=get_object_or_404(Item,pk=pk,archived=False)
        if item.character_id:
            if item.character_id!=c.id:raise PermissionDenied('Используйте свои вещи или общий запас кампании')
        else:
            access_campaign(user,item.campaign_id)
            if not c.memberships.filter(campaign_id=item.campaign_id).exists():raise PermissionDenied('Персонаж не участвует в этой кампании')
        if revision!=item.revision:raise ValueError('Запас изменился. Откройте мастерскую заново')
        return item
    production=spec['kind']=='production'
    if production and p.get('prepared') is not True:raise ValueError('Подтвердите три часа приготовления в игровом мире')
    target=None if production else stock(p['item'],p.get('item_revision'))
    if target and target.quantity<1:raise ValueError('Нет предмета для улучшения')
    field='enchantments' if spec['kind']=='enchantment' else 'upgrades'
    if target and recipe.id in target.data.get(field,[]):raise ValueError('Это улучшение уже есть на предмете')
    if spec['kind']=='enchantment':
        if target.data.get('item_type') not in enchantments.profile(recipe)['types']:raise ValueError('Чары не подходят к предмету')
    elif not production and not weaponry.compatible(target,spec):raise ValueError('Улучшение не подходит к этому оружию')
    rows=p.get('supplies',[])
    if not isinstance(rows,list) or len(rows)>30:raise ValueError('Проверьте список материалов')
    supplies=[];seen=set();dust=0;component=0
    for row in rows:
        if not isinstance(row,dict):raise ValueError('Проверьте материалы')
        item=stock(row.get('item'),row.get('revision'))
        if item.id in seen or target and item.id==target.id:raise ValueError('Материал указан дважды или совпадает с предметом')
        seen.add(item.id)
        count=amount(row.get('quantity'),item.quantity,1)
        if item.equipped:raise ValueError('Снимите предмет перед расходованием как материала')
        role=row.get('role','material')
        if role=='dust':
            if item.data.get('material_kind')!='magic_dust':raise ValueError('Выберите магическую пыль')
            dust+=count
        elif role=='component':
            if item.data.get('material_kind')!='elemental' or spec.get('element') and item.data.get('element')!=spec['element']:
                raise ValueError('Выберите компонент нужной стихии')
            component+=count
        elif role!='material':raise ValueError('Неизвестный вид материала')
        supplies.append((item,count,role))
    if spec['kind']=='enchantment' and (dust!=50 or component<1):raise ValueError('Для чар нужны 50 мер магической пыли и стихийный компонент')
    change=Change(user,'Мастерская · '+recipe.name+' · '+c.name,inputs={'character':c.id,'recipe':recipe.id,'supplies':rows})
    if target:change.watch(target)
    for item,count,role in supplies:
        change.watch(item).quantity-=count
    if production:
        result=new_item(change,c)
        result.name=recipe.name;result.quantity=1
        result.data={'item_type':'consumable','description':recipe.description,'alchemy':spec['product'],
                     'recipe_id':recipe.id,'maker_id':c.id,'preparation_hours':spec['hours']}
        change.inputs['preparation_hours']=spec['hours']
    elif target.quantity>1:
        target.quantity-=1
        result=new_item(change,target.character,target.campaign)
        result.name=target.name;result.entry_id=target.entry_id;result.data=copy.deepcopy(target.data);result.slot=target.slot;result.quantity=1
    else:result=target
    result.data=copy.deepcopy(result.data)
    if not production:result.data.setdefault(field,[]).append(recipe.id)
    change.inputs['result_item']=result.id
    change.finish()
    return {'id':result.id}

File: static/crafting.js
const materialElements=['Изначальная','Огонь','Вода','Земля','Воздух','Тьма','Свет','Молния','Холод','Природа'];
function craftLabel(i,scope){const mods=[...(i.data.enchantments||[]),...(i.data.upgrades||[])].map(id=>entry(id)?.name).filter(Boolean);return i.name+' × '+i.quantity+(mods.length?' ['+mods.join(', ')+']':'')+' · '+scope}
function craftStock(c){return [...c.items.map(i=>({...i,label:craftLabel(i,'Личный запас')})),...S.campaigns.filter(k=>k.member&&c.memberships.some(m=>m.campaign_id===k.id)).flatMap(k=>k.items.map(i=>({...i,label:craftLabel(i,k.name)})))].filter(i=>i.quantity>0)}
function upgradeFits(item,spec){const aliases={'Лук':'Луки','Арбалет':'Арбалеты','Молот':'Молоты'};return item.data.item_type==='weapon'&&[...(item.data.families||[]),...(item.data.keywords||[])].some(k=>spec.families.includes(aliases[k]||k))}
function craftFields(spec,stock){
 if(!spec)return '';
 const ench=(S.rules.enchantments||[]).find(e=>e.id===spec.id);
 const targets=stock.filter(i=>!(i.data[spec.kind==='enchantment'?'enchantments':'upgrades']||[]).includes(spec.id)&&(spec.kind==='enchantment'?ench.types.includes(i.data.item_type):upgradeFits(i,spec)));
 const targetField=`<p class="description">${esc(entry(spec.id)?.description)}</p>`+selector('craft_item','Предмет · один экземпляр',targets.map(i=>({id:i.id,name:i.label})),null);
 if(spec.kind==='enchantment'){
  let need=50;const dust=stock.filter(i=>!i.equipped&&i.data.material_kind==='magic_dust');
  return targetField+`<h3>Магическая пыль · 50 мер</h3><div class="fields">${dust.map(i=>{const n=Math.min(need,i.quantity);need-=n;return input('supply'+i.id,i.label,n,'number',`min="0" max="${i.quantity}"`) }).join('')||'<p>Добавьте магическую пыль в инвентарь как компонент.</p>'}</div>${selector('craft_component','Стихийный компонент · '+(spec.element||'по рецепту'),stock.filter(i=>!i.equipped&&i.data.material_kind==='elemental'&&(!spec.element||i.data.element===spec.element)).map(i=>({id:i.id,name:i.label})),null)}${input('component_quantity','Количество компонента',1,'number','min="1" required')}<p id="craft-budget" class="muted" aria-live="polite"></p>`;
 }
 return targetField+`<details><summary>Материалы</summary><div class="fields">${stock.filter(i=>!i.equipped).map(i=>input('supply'+i.id,i.label,0,'number',`min="0" max="${i.quantity}"`)).join('')}</div></details>`;
}
function craftingForm(c){
 const known=new Set([...c.abilities.map(a=>a.id),...(c.knowledge||[]).filter(k=>!k.archived).map(k=>k.entry_id)]);
 const crafts=[...(c.info.craft_ids||[]).map(id=>entry(id)?.name),c.info.craft];
 const recipes=(S.rules.crafting||[]).filter(r=>known.has(r.id)&&crafts.includes(r.craft));
 if(!recipes.length){modal('Мастерская','<p>Выберите ремесло в листе персонажа и изучите рецепт улучшения в разделе знаний.</p>',null);return}
 const stock=craftStock(c);
 modal('Мастерская',`${selector('craft_recipe','Рецепт',recipes,recipes[0].id)}<div id="craft-fields">${craftFields(recipes[0],stock)}</div>`,async()=>{
  const f=formObject(),spec=recipes.find(r=>r.id===num(f.craft_recipe)),item=stock.find(i=>i.id===num(f.craft_item));
  if(!spec||!item)throw Error('Выберите рецепт и предмет');
  const supplies=stock.filter(i=>num(f['supply'+i.id])>0).map(i=>({item:i.id,revision:i.revision,quantity:Number(f['supply'+i.id]),role:spec.kind==='enchantment'?'dust':'material'}));
  if(spec.kind==='enchantment'){
   const component=stock.find(i=>i.id===num(f.craft_component));if(!component)throw Error('Выберите стихийный компонент');
   supplies.push({item:component.id,revision:component.revision,quantity:Number(f.component_quantity),role:'component'});
  }
  await api({op:'craft.apply',character:c.id,recipe:spec.id,item:item.id,item_revision:item.revision,supplies});
 },'Улучшить предмет');
 const update=()=>{
  const f=formObject(),spec=recipes.find(r=>r.id===num(f.craft_recipe));let valid=!!num(f.craft_item);
  if(spec?.kind==='enchantment'){
   const dust=stock.reduce((n,i)=>n+num(f['supply'+i.id]),0),component=stock.find(i=>i.id===num(f.craft_component)),quantity=Number(f.component_quantity);
   el('craft-budget').textContent='Магическая пыль: '+dust+' / 50';
   valid=valid&&dust===50&&!!component&&Number.isInteger(quantity)&&quantity>0&&quantity<=component.quantity;
  }
  el('dialog-submit').disabled=!valid;
 };
 el('craft-fields').addEventListener('input',update);el('craft-fields').addEventListener('change',update);
 dialogBody.querySelector('[name=craft_recipe]').addEventListener('change',e=>{el('craft-fields').innerHTML=craftFields(recipes.find(r=>r.id===num(e.target.value)),stock);update()});update();
}
document.addEventListener('click',e=>{const b=e.target.closest('[data-do="craft.open"]');if(b)craftingForm(current())});
document.addEventListener('change',e=>{
 if(e.target.name==='material_kind'&&el('material-element'))el('material-element').hidden=e.target.value!=='elemental';
 if(e.target.id==='attack-mode')command({op:'attack.mode',character:current().id,mode:e.target.value}).catch(err=>toast(err.message));
});

function upgradeFields(data){return `<details ${data.upgrades?.length?'open':''}><summary>Улучшения оружейника</summary><div class="enchantment-choices">${(S.rules.crafting||[]).filter(r=>r.kind==='upgrade').map(r=>`<label><input type="checkbox" name="upgrades" value="${r.id}" ${data.upgrades?.includes(r.id)?'checked':''}><span><strong>${esc(r.name)}</strong><small>${esc(r.families.join(', '))}</small><small>${esc(entry(r.id)?.description)}</small></span></label>`).join('')}</div></details>`}
